### DISCUSSION TITLE
Can an OME-TIFF file contain BOTH a Pyramidal structure AND regular images?

### DISCUSSION THREAD

**Discussion by michael-scott-phoseon** [libvips/libvips#2458] [Q&A]
We need to have OME-TIFF files delivered to us by a 3rd party that includes  a 3 layer Pyramidal set of images, AND 2 regular images containing Slide images.  All image saved as JPEG2000.

Is this possible?  If so, how would I use libvips (or NETvips, which is what I'm currently using) to load the regular images, as well as each layer in the Pyramidal structure?

**Comment by jcupitt** [comment]
You can probably use `page=` to pull out different parts of the image. If you upload a sample I could have a look.

**Comment by michael-scott-phoseon** [comment]
I’m asking if we can supply one now.  How do I go from NetVips.Image to Page? Also, once I’ve replaced                the following with my new Metadata, how do I save it back to the ome-tiff file

                            // image-description
                            var strImageDescription = img.Get(iptcString[14]);
                            var strOmeMetadata = img.Get("image-description");


Michael Scott | Software Developer

***@***.***

7425 NE Evergreen Pkwy | Hillsboro, OR 97124<7425%20NE%20Evergreen%20Pkwy%20|%20Hillsboro,%20OR%2097124>
971.713.2665 (P) | 831.252.7268 (M)
www.phoseon.com<http://www.phoseon.com/>

**Reply by jcupitt** [reply]
You should ask on the netvips issue tracker for a C# solution, but in python you'd just use `write_to_file` to write the modified image back to storage.

`page` is an argument to `new_from_file` that lets you select which image page to load. You'll need to examine the xml to see which page(s) you need to read.

**Comment by michael-scott-phoseon** [comment]
I’m close – but saving the file after injecting new metadata knocks the file down to a fraction of the size of the original file.  Ideas of where I went wrong?

I'm trying to inject additional information into the OME-TIFF metadata of a multi-image ome-tiff file. This sample tiff file contains 2 regular images, and a set of pyramidal images. I'm using the following code, but the resulting file is over HALF the size of the other. any ideas what I'm doing wrong:

The OME-TIFF file I am working with is here:

https://downloads.openmicroscopy.org/images/OME-TIFF/2016-06/sub-resolutions/Fluorescence/LuCa-7color_Scan1.ome.tiff

And the code I am using is below. Suggestions of what I am doing wrong? Ultimately, I want to save this to the same file, but saving the changed file to another one is OK for now.

```C#
                    using (var img = NetVips.Image.NewFromFile(slide))
                    {
                        var strImageDescription = img.Get("image-description");

                        XmlSerializer Serializer = new XmlSerializer((typeof(OME)));
                        using (TextReader tr = new StringReader(strImageDescription.ToString()))
                        {
                            metadata = (OME)Serializer.Deserialize(tr);
                        }

                        // Convert to List
                        List<object> lstItems = new List<object>(metadata.Items.ToList());

                        // If Experimenter doesn't exist, add it
                        if (lstItems.Any(i => i.GetType() == typeof(Experimenter)) == false)
                        {
                            lstItems.Add((object)new Experimenter()
                                {
                                    FirstName = "Fred", LastName = "Flintstone,
                                    Email = ***@***.******@***.***>, Institution = Constants.CompanyName,
                                    UserName = Environment.UserName
                                });
                        }

                        // Get or create StructuredAnnotations block
                        StructuredAnnotations structuredAnnotations =
                            (StructuredAnnotations)lstItems.FirstOrDefault(i => i.GetType() == typeof(StructuredAnnotations));

                        // We need the Index, so we can delete / replace it later
                        int structuralAnnotationsIdx = lstItems.FindIndex(i => i.GetType() == typeof(StructuredAnnotations));

                        if (structuredAnnotations == null)
                           structuredAnnotations = new StructuredAnnotations();

                        // Get and / or create list of TagAnnotation Annotations
                        List<Annotation> lstAnnotations = new List<Annotation>();
                        if (structuredAnnotations.Items?.Length > 0)
                            lstAnnotations = structuredAnnotations.Items.ToList();

                        if (lstAnnotations.Any(i => i.GetType() == typeof(TagAnnotation)) == false)
                        {
                            lstAnnotations.Add(new TagAnnotation() { ID = "Ann:1", Description = "RosyBrown", Value = "Note 1" });
                            lstAnnotations.Add(new TagAnnotation() { ID = "Ann:2", Description = "RoyalBlue", Value = "Note 2" });
                            lstAnnotations.Add(new TagAnnotation() { ID = "Ann:3", Description = "SpringGreen", Value = "Note 3" });
                        }

                        // Add TimestampAnnotation
                        lstAnnotations.Add(new TimestampAnnotation()
                                { Description = $"UserName: {Environment.UserName}", Value = DateTime.Now });

                        // Delete old Structured Annotations
                        if (structuralAnnotationsIdx >= 0)
                            lstItems.RemoveAt(structuralAnnotationsIdx);

                        // Add then add new one back
                        structuredAnnotations.Items = lstAnnotations.ToArray();
                        lstItems.Add((object)structuredAnnotations);

                        // Now update Metadata Items
                        metadata.Items = lstItems.ToArray();

                        // And convert it to XML string
                        string newMetadata = string.Empty;
                        XmlSerializer serializer = new XmlSerializer(typeof(OME));
                        using (StringWriter sw = new StringWriter())
                        {
                            serializer.Serialize(sw, metadata);
                            newMetadata = sw.ToString();
                        }

                        using var mutated = img.Mutate(mutable =>
                        {
                            foreach (var field in img.GetFields())
                            {
                                if (field == "image-description")
                                    mutable.Set(GValue.GStrType, nameof(newMetadata), newMetadata);
                            }
                        });

                        img.WriteToFile("C:\\abcd.OME.TIFF");
                    }
```

**Comment by jcupitt** [comment]
The file `LuCa-7color_Scan1.ome.tiff` contains 8 pages. You can see the structure with `tifffinfo` (I've trimmed the `IMAGEDESCRIPTION` tags to make it easier to read):

```
$ tiffinfo LuCa-7color_Scan1.ome.tiff 
TIFF Directory at offset 0x10 (16)
  Image Width: 24960 Image Length: 34560
  Tile Width: 512 Tile Length: 512
  Resolution: 20080.5, 20080.5 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: LZW
  Photometric Interpretation: min-is-black
  Samples/Pixel: 1
  Planar Configuration: single image plane
  SubIFD Offsets: 1533039599 1920856308 2021666749 2052152209 2060362679
  ImageDescription: <?xml version="1.0" encoding="UTF-8"?>...
TIFF Directory at offset 0x14d40236 (349438518)
  Image Width: 24960 Image Length: 34560
  Tile Width: 512 Tile Length: 512
  Resolution: 20080.5, 20080.5 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: LZW
  Photometric Interpretation: min-is-black
  Samples/Pixel: 1
  Planar Configuration: single image plane
  SubIFD Offsets: 1619859285 1943702908 2028625295 2054020223 2060864623
  ImageDescription: ImageJ= ...
TIFF Directory at offset 0x26fcd232 (654103090)
  Image Width: 24960 Image Length: 34560
  Tile Width: 512 Tile Length: 512
  Resolution: 20080.5, 20080.5 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: LZW
  Photometric Interpretation: min-is-black
  Samples/Pixel: 1
  Planar Configuration: single image plane
  SubIFD Offsets: 1697674728 1963786396 2034630073 2055633452 2061309554
  ImageDescription: ImageJ= ...
TIFF Directory at offset 0x34bc22ea (884744938)
  Image Width: 24960 Image Length: 34560
  Tile Width: 512 Tile Length: 512
  Resolution: 20080.5, 20080.5 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: LZW
  Photometric Interpretation: min-is-black
  Samples/Pixel: 1
  Planar Configuration: single image plane
  SubIFD Offsets: 1754506020 1978124093 2039120613 2056875689 2061661251
  ImageDescription: ImageJ= ...
TIFF Directory at offset 0x48985603 (1217943043)
  Image Width: 24960 Image Length: 34560
  Tile Width: 512 Tile Length: 512
  Resolution: 20080.5, 20080.5 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: LZW
  Photometric Interpretation: min-is-black
  Samples/Pixel: 1
  Planar Configuration: single image plane
  SubIFD Offsets: 1840565049 2000694891 2045814865 2058654171 2062141161
  ImageDescription: ImageJ= ...
TIFF Directory at offset 0x7af0e05a (2062606426)
  Image Width: 195 Image Length: 270
  Resolution: 156.878, 156.878 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: LZW
  Photometric Interpretation: RGB color
  Samples/Pixel: 3
  Rows/Strip: 1
  Planar Configuration: separate image planes
  ImageDescription: ImageJ= ...
TIFF Directory at offset 0x7af2b0b6 (2062725302)
  Image Width: 2090 Image Length: 4036
  Resolution: 720.689, 720.689 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: LZW
  Photometric Interpretation: RGB color
  Samples/Pixel: 3
  Rows/Strip: 1
  Planar Configuration: separate image planes
  ImageDescription: ImageJ= ...
TIFF Directory at offset 0x7be87477 (2078831735)
  Image Width: 522 Image Length: 450
  Resolution: 180.172, 180.172 pixels/cm
  Bits/Sample: 8
  Sample Format: unsigned integer
  Compression Scheme: LZW
  Photometric Interpretation: RGB color
  Samples/Pixel: 3
  Rows/Strip: 1
  Planar Configuration: separate image planes
  ImageDescription: ImageJ= ...
```

So the 8 pages are arranged like this:

1. 24960 x 34560 pixels, 1 8-bit band, 512x512 tiles, LZW compression, pyramid, OME metadata
2. to 5. Same, but ImageJ metadata
6. 195 x 270 pixel, RGB, LZW compression, ImageJ metadata
7. 2090 x 4036 RGB, LZW compression, ImageJ metadata
8. 522 x 450, RGB, LZW compression, ImageJ metadata

ie. 1-5 is the 5 channel pyramid, 6 is an RGB preview, 7 is the macro, and 8 is the label. The OME metadata has this info.

If you want to replicate this structure exactly, just with changed OME metadata, you'll need to be able to parse the file and extract all this information. I think LZW compression is a little unusual, most OME-TIFFs will use high-quality JPEG instead.

I would copy each page separately, then use `tiffcp` at the end to join the 8 pages back into a single file.

**Comment by michael-scott-phoseon** [comment]
John,
LZZW
The file I provided the link for is just the sample I’m using so I can work on images with both the pyramid structure, and regular images.  I have no loyalty to that file.  I DO know we will be using JPEG2000, not LZW compression though.  We just don’t have any with both images types yet.

I do want to exactly replicate the entire file, both images and Metadata.  I thought that I was preserving the original Metadata, and only injected <Experimenter>, then ADDED TO the <StructuredAnnotations> section (or created one, if it didn’t exist).  The following code, when looking at the contents, contained all the original Metadata, but included my additions

                            // Now update Metadata Items
                            metadata.Items = lstItems.ToArray();

Then this code stuffed the new metadata back into the file:

                            // And convert it to XML string
                            string newMetadata = string.Empty;
                            XmlSerializer serializer = new XmlSerializer(typeof(OME));
                            using (StringWriter sw = new StringWriter())
                            {
                                serializer.Serialize(sw, metadata);
                                newMetadata = sw.ToString();
                            }

                            using var mutated = img.Mutate(mutable =>
                            {
                                foreach (var field in img.GetFields())
                                {
                                    if (field == "image-description")
                                        mutable.Set(GValue.GStrType, nameof(newMetadata), newMetadata);
                                }
                            });

So what did I misunderstand?  I will verify that again today

Also, I’d like to save this back to the SAME file once I know I’m doing it right, instead of the following:

img.WriteToFile("C:\\abcd.OME.TIFF");

How best to do that?
NOTICE: This communication may contain privileged or other confidential information. If you have received it in error, please advise the sender by reply email and immediately delete the message and any attachments without copying or disclosing the contents. Thank you.

**Reply by jcupitt** [reply]
Sorry, I don't really understand what you're doing. Are you loading all the pages? Are you using `tiffcp`? You'd need to post a complete, runnable program I can try. Kleis (the NetVips maintainer) might be a better person to ask if the question is specifically about C#.

I find python very handy for prototyping, I don't know if that's an option for you.

**Reply by jcupitt** [reply]
I'd guess you can just write:

```C#
                            using var mutated = img.Mutate(mutable =>
                            {
                                mutable.Set(GValue.GStrType, "image-description", newMetadata);
                            });
```

But it's just a guess.

**Reply by jcupitt** [reply]
Oh, and then you need to write the mutated image. At the moment, you're writing the original. Try:

```C#
                        mutated.WriteToFile("C:\\abcd.OME.TIFF");
```

**Comment by michael-scott-phoseon** [comment]
Assuming Compression was Jp2k, instead of LZW, should this work?

                       using (var img = NetVips.Image.NewFromFile(slide))
                        {
                            int? ifdThumbnail = 5;
                            int? ifdOverview = 6;
                            int? ifdLabel = 7;

                            var thumbnail = NetVips.Image.Jp2kload(slide, ifdThumbnail, access: Enums.Access.Random);
                            var overview = NetVips.Image.Jp2kload(slide, ifdOverview, access: Enums.Access.Random);
                            var label = NetVips.Image.Jp2kload(slide, ifdLabel, access: Enums.Access.Random);

Wow would I load these 3 images since it’s LXW  Howe about the pyramidal image?

I will write a small program for you tomorrow that demonstrates what I’m trying to do.  But right now, all I want to do is take this image:

https://downloads.openmicroscopy.org/images/OME-TIFF/2016-06/sub-resolutions/Fluorescence/LuCa-7color_Scan1.ome.tiff

And make an exact copy of it with <Experimenter> and a few <TagAnnotations> stuffed into <StructuredAnnotations>  I also want to be able to stuff ALL 4 of these images into a Windows Form (actually, a WPF Window).
NOTICE: This communication may contain privileged or other confidential information. If you have received it in error, please advise the sender by reply email and immediately delete the message and any attachments without copying or disclosing the contents. Thank you.

**Reply by jcupitt** [reply]
Your program needs to look something like this (using python for convenience, but the C# API is the same):

```python
import pyvips

# this will read the header from page 0 and not read any pixels
image = pyvips.Image.tiffload(filename)
ome_metadata = image.get("image-description")

# analyze metadata to find the number and type of the various pages
# update metadata with new stuff

# copy each page to a new tiff file
for i in range(n_pages):
    image = pyvips.Image.tiffload(filename, page=i)

    # update the metadata on page 0
    if i == 0:
        # this is the py equivalent of the mutate feature
        image = image.copy()
        image.set("image-description", new_ome_metadata)

    # you'll need to create the settings arrays in your metadata analysis above
    image.tiffsave(f"page-{i:03}.tif",
        compression=compression_setting[i],
        tile=tile_setting[i],
        tile_width=tile_width_setting[i],
        tile_height=tile_height_setting[i],
        pyramid=pyramid_setting[i],
        subifd=subifd_setting[i])

# combine the N tiff images into a single TIFF output
os.system("tiffcp page-*.tif final.tif")
```

**Comment by michael-scott-phoseon** [comment]
Could you show the following in C++ or c# using Mutate:  I’d rather not saving all these images to disk, as these tiffs will be huge, so I’d like to keep file I/O to a minimum


    # you'll need to create the settings arrays in your metadata analysis above

    image.tiffsave(f"page-{i:03}.tif",

        compression=compression_setting[i],

        tile=tile_setting[i],

        tile_width=tile_width_setting[i],

        tile_height=tile_height_setting[i],

        pyramid=pyramid_setting[i],

        subifd=subifd_setting[i])



# combine the N tiff images into a single TIFF output

os.system("tiffcp page-*.tif final.tif")

**Reply by jcupitt** [reply]
I'm afraid it's not possible to modify a TIFF file on disc, the format does not allow it. You're going to have to read the whole thing in and write it out again.

**Comment by michael-scott-phoseon** [comment]
I was referring to taking an one tiff file  containing 8 images, modifying it’s metadata, saving all 8 images to disk as 8 files, then combining all 8 files back into one file again.  The pyramidal images may be 4 gig so the fewer accesses to the larger images the better

Isn’t that what the last 2 sections of your python code does?  I’m afraid I don’t know python, so maybe I misunderstood your code.

I take it that Image.Copy doesn’t copy all the images? Nor does Mutableimage?

--
Michael G Scott
Sent from my iPhone


On Oct 6, 2021, at 8:54 PM, John Cupitt ***@***.***> wrote:

﻿

I'm afraid it's not possible to modify a TIFF file on disc, the format does not allow it. You're going to have to read the whole thing in and write it out again.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/libvips/libvips/discussions/2458#discussioncomment-1440124>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ARZ2TVZWVBBKM7T433ZXOPLUFUKYJANCNFSM5FFT2VCA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

NOTICE: This communication may contain privileged or other confidential information. If you have received it in error, please advise the sender by reply email and immediately delete the message and any attachments without copying or disclosing the contents. Thank you.

**Reply by jcupitt** [reply]
Yes, `image1 = image2.copy()` just copies a pointer, no pixels are copied. `MutableImage` is the same: it's just a wrapper around a copy.

The `tiffcp` is unfortunate, but unavoidable with NetVips. It's pretty quick, on the plus side.
